### PR TITLE
Improve inventory tests; few fixes

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by trasher** [comment]
Certainly to be reviewed per commit

File: tests/functionnal/Glpi/Inventory/Assets/Controller.php [hunk 1]
```diff
@@ -130,4 +130,180 @@ public function testHandle() {
...
+
+      //controllers present in the inventory source are now dynamic
+      $controllers = $item_control->find(['itemtype' => 'Computer', 'items_id' => $computers_id, 'is_dynamic' => 1]);
+      $this->integer(count($controllers))->isIdenticalTo(2);
+
+      //controller not present in the inventory is still not dynamic
+      $controllers = $item_control->find(['itemtype' => 'Computer', 'items_id' => $computers_id, 'is_dynamic' => 0]);
+      $this->integer(count($controllers))->isIdenticalTo(1);
+
+      //Redo inventory, but with removed "NVMe SSD" processor
```
Comment (g-bougard) [inline tests/functionnal/Glpi/Inventory/Assets/Controller.php:263]: You removed a controller, not a "processor" ;-)

**Review by g-bougard — APPROVED** [review]
I only see one comment to be fixed.

File: inc/item_operatingsystem.class.php [hunk 2]
```diff
@@ -41,6 +41,7 @@ class Item_OperatingSystem extends CommonDBRelation {
    static public $itemtype_2 = 'itemtype';
    static public $items_id_2 = 'items_id';
    static public $checkItem_1_Rights = self::DONT_CHECK_ITEM_RIGHTS;
+   static public $checkItem_2_Rights = self::DONT_CHECK_ITEM_RIGHTS;
```
Comment (cedric-anne) [inline inc/item_operatingsystem.class.php:44]: I am not really confident with this.

File: tests/functionnal/Glpi/Inventory/Assets/Peripheral.php [hunk 3]
```diff
@@ -110,4 +110,179 @@ public function testHandle() {
...
+      ]);
+      $this->integer($item_periph_3_id)->isGreaterThan(0);
+
+      $periphs = $item_periph->find(['itemtype' => \Peripheral::class, 'computers_id' => $computers_id]);
+      $this->integer(count($periphs))->isIdenticalTo(3);
+      foreach ($periphs as $p) {
+         $this->variable($p['is_dynamic'])->isEqualTo(0);
+      }
+
+      //computer inventory knows only "Fingerprint" and "Smatrtcard reader" peripherals
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/Peripheral.php:226]:
```suggestion
      //computer inventory knows only "Fingerprint" and "Smartcard reader" peripherals
```

File: tests/functionnal/Glpi/Inventory/Assets/Peripheral.php [hunk 4]
```diff
@@ -110,4 +110,179 @@ public function testHandle() {
...
+
+      //peripherals present in the inventory source are now dynamic
+      $periphs = $item_periph->find(['itemtype' => \Peripheral::class, 'computers_id' => $computers_id, 'is_dynamic' => 1]);
+      $this->integer(count($periphs))->isIdenticalTo(2);
+
+      //peripheral not present in the inventory is still not dynamic
+      $periphs = $item_periph->find(['itemtype' => \Peripheral::class, 'computers_id' => $computers_id, 'is_dynamic' => 0]);
+      $this->integer(count($periphs))->isIdenticalTo(1);
+
+      //Redo inventory, but with removed "Smatrtcard reader" peripheral
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/Peripheral.php:245]:
```suggestion
      //Redo inventory, but with removed "Smartcard reader" peripheral
```

File: tests/functionnal/Glpi/Inventory/Assets/OperatingSystem.php [hunk 5]
```diff
@@ -759,4 +758,118 @@ public function testHandle() {
...
+         'operatingsystems_id' => $os_id
+      ]);
+      $this->integer($cos_id)->isGreaterThan(0);
+
+      $this->doInventory($xml_source, true);
+
+      $list = $os->find();
+      $this->integer(count($list))->isIdenticalTo(1);
+
+      //check tha tOS is linked to computer, and is now dynamic
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/OperatingSystem.php:823]:
```suggestion
      //check that OS is linked to computer, and is now dynamic
```

File: tests/functionnal/Glpi/Inventory/Assets/OperatingSystem.php [hunk 6]
```diff
@@ -759,4 +758,118 @@ public function testHandle() {
...
+  </CONTENT>
+  <DEVICEID>test-pc002</DEVICEID>
+  <QUERY>INVENTORY</QUERY>
+</REQUEST>";
+
+      //create manually a computer, with an operating system
+      $computers_id = $computer->add([
+         'name'   => 'pc002',
+         'serial' => 'ggheb7ne7',
+         'entities_id' => 0
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/OperatingSystem.php:802]: Maybe your right issue is linked to this. Your computer is maybe not visible in sub-entities, and TU user uses `_test_root_entity` by default.

```suggestion
         'entities_id' => getItemByTypeName('Entity', '_test_root_entity', true)
```

File: inc/item_operatingsystem.class.php (hunk 2 above)
Comment (trasher) [reply inc/item_operatingsystem.class.php]: I've reverted that change; I've made a small change on inventory regarding entity stored in session; and that fixed it.

File: inc/inventory/asset/mainasset.class.php [hunk 7]
```diff
@@ -657,10 +660,18 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0) {
...
 
-      if ($orig_entity !== null) {
-         $_SESSION['glpiactive_entity'] = $orig_entity;
+      //Restore entities in session
+      if ($orig_glpiactive_entity !== null) {
+         $_SESSION['glpiactive_entity'] = $orig_glpiactive_entity;
+      }
+
+      if ($orig_glpiactiveentities !== null) {
+         $_SESSION['glpiactiveentities'] = $orig_glpiactive_entity;
```
Comment (cedric-anne) [inline inc/inventory/asset/mainasset.class.php:669]:
```suggestion
         $_SESSION['glpiactiveentities'] = $orig_glpiactiveentities;
```

File: inc/inventory/asset/mainasset.class.php [hunk 8]
```diff
@@ -657,10 +660,18 @@ public function rulepassed($items_id, $itemtype, $rules_id, $ports_id = 0) {
...
       }
 
+      if ($orig_glpiactiveentities_string !== null) {
+         $_SESSION['glpiactiveentities_string'] = $orig_glpiactive_entity;
```
Comment (cedric-anne) [inline inc/inventory/asset/mainasset.class.php:673]:
```suggestion
         $_SESSION['glpiactiveentities_string'] = $orig_glpiactiveentities_string;
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: tests/functionnal/Glpi/Inventory/Assets/Monitor.php [hunk 9]
```diff
@@ -170,4 +170,127 @@ public function testHandle() {
...
+  </CONTENT>
+  <DEVICEID>test-pc003</DEVICEID>
+  <QUERY>INVENTORY</QUERY>
+</REQUEST>";
+
+      //computer inventory with one monitor
+      $inventory = $this->doInventory($xml_source_2, true);
+
+      $computers_2_id = $inventory->getItem()->fields['id'];
+      $this->integer($computers_id)->isGreaterThan(0);
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/Monitor.php:259]:
```suggestion
      $this->integer($computers_2_id)->isGreaterThan(0);
```

File: tests/functionnal/Glpi/Inventory/Assets/Peripheral.php [hunk 10]
```diff
@@ -285,4 +285,129 @@ public function testInventoryUpdate() {
...
+  </CONTENT>
+  <DEVICEID>test-pc003</DEVICEID>
+  <QUERY>INVENTORY</QUERY>
+</REQUEST>";
+
+      //computer inventory with one peripheral
+      $inventory = $this->doInventory($xml_source_2, true);
+
+      $computers_2_id = $inventory->getItem()->fields['id'];
+      $this->integer($computers_id)->isGreaterThan(0);
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/Peripheral.php:376]:
```suggestion
      $this->integer($computers_2_id)->isGreaterThan(0);
```

File: tests/functionnal/Glpi/Inventory/Assets/Printer.php [hunk 11]
```diff
@@ -183,4 +183,137 @@ public function testSnmpPrinter() {
...
+  </CONTENT>
+  <DEVICEID>test-pc003</DEVICEID>
+  <QUERY>INVENTORY</QUERY>
+</REQUEST>";
+
+      //computer inventory with one printer
+      $inventory = $this->doInventory($xml_source_2, true);
+
+      $computers_2_id = $inventory->getItem()->fields['id'];
+      $this->integer($computers_id)->isGreaterThan(0);
```
Comment (cedric-anne) [inline tests/functionnal/Glpi/Inventory/Assets/Printer.php:282]:
```suggestion
      $this->integer($computers_2_id)->isGreaterThan(0);
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**